src/coreclr/jit/codegenxarch.cpp

```diff
@@ -4842,19 +4842,11 @@ void CodeGen::genCodeForShiftLong(GenTree* tree)
 
     unsigned int count = (unsigned int)shiftBy->AsIntConCommon()->IconValue();
 
-    regNumber regResult = (oper == GT_LSH_HI) ? regHi : regLo;
+    regNumber tgtReg = tree->GetRegNum();
+    assert(regLo != tgtReg);
 
-    inst_Mov(targetType, tree->GetRegNum(), regResult, /* canSkip */ true);
-
-    if (oper == GT_LSH_HI)
-    {
-        inst_RV_RV_IV(ins, emitTypeSize(targetType), tree->GetRegNum(), regLo, count);
-    }
-    else
-    {
-        assert(oper == GT_RSH_LO);
-        inst_RV_RV_IV(ins, emitTypeSize(targetType), tree->GetRegNum(), regHi, count);
-    }
+    inst_Mov(targetType, tgtReg, regHi, /* canSkip */ true);
+    inst_RV_RV_IV(ins, emitTypeSize(targetType), tgtReg, regLo, count);
```

Comment:
The actual operands here are all swapped around. This is generating `result = (upper << count) | (lower >>> (32 - count))` for example.

For `DecomposeRotate` when `count < 32`, this ends up being the following (which is correct):
```
result.lower = (value.lower << count) | (value.upper >>> (32 - count))
result.upper = (value.upper << count) | (value.lower >>> (32 - count))
```

This is because `DecomposeRotate` passes it in as `hiCopy, loOp1` and then as `loCopy, hiOp1`.

Ideally this would be swapped so that its passed in as `loOp1, hiCopy` then `hiOp1, loCopy` with codegen doing:
```cpp
    regNumber tgtReg = tree->GetRegNum();
    assert(regHi != tgtReg);

    inst_Mov(targetType, tgtReg, regLo, /* canSkip */ true);
    inst_RV_RV_IV(ins, emitTypeSize(targetType), tgtReg, regHi, count);
```

However, swapping these + LSRA was leading to failures and I didn't bother investigating further since its already working "as is".

Reply:
> However, swapping these + LSRA was leading to failures

I wonder if that is because you need to change LIR emitted for the ROR (the order of args/clones)

Reply:
I had thought I had gotten that all when testing, but it's possible I missed something when trying. It's likely not worth extra effort to solve though since this is all working now.